src/pfe/portal/modules/Templates.js

```diff
@@ -546,3 +545,113 @@ function isTemplateSummary(obj) {
+  }
+  catch (err) {
+    throw new TemplateError('ADD_TO_PROVIDER_FAILURE', repo.url, err.message);
+  }
+  return providers;
+}
+
+async function removeRepositoryFromProviders(repo, providers) {
+  const promises = [];
+  for (const provider of Object.values(providers)) {
+    if (typeof provider.canHandle === 'function') {
+      // make a new copy to for each provider to be invoked with
+      // in case any provider modifies it (which they shouldn't do)
+      const copy = Object.assign({}, repo);
+      if (provider.canHandle(copy) && typeof provider.removeRepository === 'function')
+        promises.push(provider.removeRepository(copy));
+    }
+  }
+  await Promise.all(promises);
+  return providers;
+}
+
+async function performOperationsOnRepositoryList(requestedOperations, repositoryList) {
+  // const newRepositoryList = cwUtils.deepClone(repositoryList);
+  const newRepositoryList = [];
+  const promiseList = requestedOperations.map(async operation => {
+    const { operationResult, updatedRepo } = await performOperationOnRepository(operation, repositoryList);
+    if (updatedRepo) {
+      // Update the repository list with the new repository
+      newRepositoryList.push(updatedRepo);
+    }
+    return operationResult;
+  });
+  const operationResults = await Promise.all(promiseList);
+  return { operationResults, newRepositoryList }
+}
+
+async function performOperationOnRepository(operation, repositoryList) {
+  const { op, url, value } = operation;
+  const repo = repositoryList.find(repo => repo.url === url);
+  let resultObject = {};
+  if (op === 'enable') {
+    resultObject = await enableOrDisableRepository({ value }, repo);
+  }
+
+  const { response: operationResult, updatedRepo } = resultObject;
+  operationResult.requestedOperation = operation;
+  return { operationResult, updatedRepo };
+}
+
+/**
+ * @param {JSON} { url (URL of template repo to enable or disable), value (true|false)}
+ * @returns {JSON} { response: { status, error (optional) }, updatedRepo (optional) }
+ */
+function enableOrDisableRepository({ value }, repo) {
+  if (!repo) {
+    return {
+      response: {
+        status: 404,
+        error: 'Unknown repository URL',
+      }
+    }
+  }
+  try {
+    repo.enabled = (value === 'true' || value === true);
+    return {
+      response: {
+        status: 200,
+      },
+      updatedRepo: repo,
+    }
+  } catch (error) {
+    return {
+      response: {
+        status: 500,
+        error: error.message,
+      }
+    }
+  }
+}
+
+// Updates repositories in the repositoryList with the updated values
+// New repositories will not be added
+// Repositories with no updates will be kept with no changes
+function updateRepositoryList(currentRepositoryList, updatedRepositories) {
+  const newRepositoryList = currentRepositoryList.map(oldRepo => {
+    const modifiedRepo = updatedRepositories.find(modifiedRepo => modifiedRepo.url === oldRepo.url);
+    if (modifiedRepo) {
```

Comment:
as above, think you could do `return modifiedRepo || oldRepo`

Reply:
🔥